### PR TITLE
#11527 - Webchat Accessibility

### PR DESCRIPTION
**Pull request by sushmi21**
This PR adds aria attributes, roles, visible focus styles and alt texts (if one exist) to improve accessibility.

Prerequisites:

1. Install WAVE  (chrome extension)  - Quick tool to evaluate the accessibility of athe page
2. Install ChromeVox (chrome extension) - Screen Reader / NVDA - Desktop App for Windows - for better results 
3. Create a flow that includes all possible interaction elements (like gallery elements with button, list, audio and video attachments)

Testing:

Please test this against this PR for the UI: [https://cognigy.visualstudio.com/Cognigy.AI/_git/cognigy/pullrequest/3761](url)

**-Alternate Text**

1. In order to add an alternate text, create a say node in our ui.
2. Configure the node to add and image. After adding the image url, add an alternate text in the new input field 'Alternate Text'
3. You will now be able to get the alternate text in the webchat, if you inspect the image.
4. Similarly, you will be able to add alternate text for video and audio (both in default and webchat tab)

**- Keyboard Navigation**

1. Use just your keyboard to interact with bot
3. All focusable elements should be able to receive a focus on Tab / Shift Tab
2. You should be able to visually identify the focused element (Change in color or shadow)
3. Once the whole chat history panel is focused, you should be able to scroll through the chat using Up / Down Arrows

**- Screen Reader Compatability**

1. Now, enable your ChromeVox screen reader (or any other)
2. You should be able to hear the aria-label or label of the buttons when they receive focus
3. When you send/ receive a message, you should be able to differentiate between different source of the message. (Ex, when receiving a message, you should hear 'Bot say bla bla bla'
4. When receiving a message, the new messages should be read out. (When multiple messages are received at once, ChromeVox  starts reading the new message before completing the previous one. I think this will not be the case in good screen readers)
5. When the whole chat content receives focus, all messages should be read out
6. While placing the cursor on a single message item, the content of that particular message has to be read out. (You should hear the difference between audio / video / image / text content)
7. For lists, you should hear the screen reader specifying it as list.
8. When there is an alt text is present for image / audio / video, it should be read out by the screen reader. 
9. In some cases, the alt text will be empty string, so that the decorative or uninformative image will be ignored by the screen reader in the absence of alt text (Ex, image inside quick reply button,image in list item).
10. In other cases, where we have large images, a predefined alt text should be read out. (Ex, Image attachment with no provided alt text). This is necessary because, we use `div` tags with `background-image` style to display images.

**- Testing using WAVE**

1. Once you install WAVE in chrome, you can right click on any page and select 'WAVE this page'
2. See if there are no errors (other than doc title and lang)

**- Testing using Accessibility tab in firefox dev tools**

1. Inspect the document structure in Accessibility tab. 
2. Check if there are no errors

Known Issue: 

 **Accessibility of date picker:** The calender can be opened or closed using keyboard controls, but specific year/ month/ date cannot be selected using keyboard.

### CODE REVIEW

**Review by kwinto — COMMENTED** [review]
- No messages are being announced in VoiceOver in ios. Do you have access to any ios device to check what may be the issue?

- The message previews (when Webchat is closed) are not being announced in general.

File: src/webchat-ui/components/WebchatUI.tsx [hunk 1]
```diff
@@ -342,14 +342,31 @@ export class WebchatUI extends React.PureComponent<React.HTMLProps<HTMLDivElemen
...
+		const showDisconnectOverlay = enableConnectionStatusIndicator && !connected && hadConnection;
+		
+		const openChatAriaLabel = () => {
+			switch (unseenMessages.length) {
+				case 0:
+					return "Chat";
+				case 1:
+					return "One unread message in chat";		
+				default:
+					return `${unseenMessages.length} unread messages in chat`;
```
[the remark this answers is not in the document]
Comment (sushmi21) [reply src/webchat-ui/components/WebchatUI.tsx:354]: Unfortunately, I do not have any ios devices

[the remark this answers is not in the document]
Comment (sushmi21) [reply src/webchat-ui/components/WebchatUI.tsx]: Can you tell me which browser you used with the VoiceOver? Safari or any other? It should work with Safari for sure. Can you check that for me, if you haven't already?

[the remark this answers is not in the document]
Comment (sushmi21) [reply src/webchat-ui/components/WebchatUI.tsx]: And, regarding the messages not being announced: The message will not be read out loud when the chat window is closed or the user is on a different tab. But, you will hear a message tone for every incoming message when the chat is closed or when the user is on a different tab.

[the remark this answers is not in the document]
Comment (kwinto) [reply src/webchat-ui/components/WebchatUI.tsx]: I mean the announcement of the preview you get if you enable `enableUnreadMessagePreview` setting

[the remark this answers is not in the document]
Comment (kwinto) [reply src/webchat-ui/components/WebchatUI.tsx]: >Can you tell me which browser you used with the VoiceOver? Safari or any other? It should work with Safari for sure. Can you check that for me, if you haven't already? @kwinto . Also, please check your VoiceOver settings. In the settings, please check 'Enable Live Region', if it is not already.

I think you reference the mac version of VoiceOver here. I have checked only the one built-in ios (there is no live region related settings there). 

I will look into VoiceOver issue.

[the remark this answers is not in the document]
Comment (sushmi21) [reply src/webchat-ui/components/WebchatUI.tsx]: > I mean the announcement of the preview you get if you enable `enableUnreadMessagePreview` setting

Fixed!

**Comment by pedily** [comment]
The focus highlighting should not affect positioning / layout of elements.
For media templates, the border will move the content a bit. Can we change that to a "highlighted" shadow? Those don't affect the layout
ezgif-1-313f5c48f7c1

File: src/plugins/messenger/MessengerPreview/components/MessengerFrame.tsx [hunk 2]
```diff
@@ -7,7 +7,23 @@ export const getMessengerFrame = ({ React, styled }: MessagePluginFactoryProps)
...
         backgroundColor: 'hsl(0, 0%, 95%)',
         color: 'hsla(0, 0%, 0%, .8)',
-        boxShadow: theme.messageShadow
+        boxShadow: theme.messageShadow,
+        "& div": {
+			"& audio": {
+                borderRadius: theme.unitSize * 2,
+				"&:focus": {
+                    outline: 'none',
+                    borderLeft: `2px solid ${theme.primaryColor}`,
```
Comment (pedily) [inline src/plugins/messenger/MessengerPreview/components/MessengerFrame.tsx:16]: this is affecting the layout

File: src/plugins/messenger/MessengerPreview/components/MessengerFrame.tsx [hunk 3]
```diff
@@ -7,7 +7,23 @@ export const getMessengerFrame = ({ React, styled }: MessagePluginFactoryProps)
...
+                },
+            },
+            "& video": {
+                borderRadius: theme.unitSize * 2,
+				"&:focus": {
+                    outline: 'none',
+                    borderLeft: `2px solid ${theme.primaryColor}`,
```
Comment (pedily) [inline src/plugins/messenger/MessengerPreview/components/MessengerFrame.tsx:23]: this is affecting the layout

**Comment by pedily** [comment]
The WAVE tool is complaining about the white text on red background for the "unread message indicator"

**Review by pedily — COMMENTED** [review]
The WAVE tool is complaining about the white text on red background for the "unread message indicator".
You can trigger that by using webchat.sendMessage('some text') in your embed code without opening the webchat

**Comment by pedily** [comment]
WAVE is also complaining about the low contrast of the "powered by cognigy" banner, but i guess that's okay since it's meant to be more of a watermark than content

File: src/plugins/messenger/MessengerPreview/components/MessengerFrame.tsx (hunk 2 above)
Comment (sushmi21) [reply src/plugins/messenger/MessengerPreview/components/MessengerFrame.tsx]: fixed

File: src/plugins/messenger/MessengerPreview/components/MessengerFrame.tsx (hunk 3 above)
Comment (sushmi21) [reply src/plugins/messenger/MessengerPreview/components/MessengerFrame.tsx]: fixed

**Comment by sushmi21** [comment]
> The WAVE tool is complaining about the white text on red background for the "unread message indicator".
> You can trigger that by using webchat.sendMessage('some text') in your embed code without opening the webchat

Fixed

**Comment by sushmi21** [comment]
> WAVE is also complaining about the low contrast of the "powered by cognigy" banner, but i guess that's okay since it's meant to be more of a watermark than content

It is fine for this text

**Comment by kwinto** [comment]
Gallery next/previous buttons need an indication of focused state.

**Comment by sushmi21** [comment]
> Gallery next/previous buttons need an indication of focused state.

Initially, the button appeared only on hover. Now, the button will appear on receiving focus. Although there is no outline/ border/ shadow to the button, the appearance of the button itself implies that the button is the element that is currently receiving focus.

**Review by pedily — APPROVED** [review]

**Review by kwinto — APPROVED** [review]

**Pull request merged**